File: RokidSDKDemoJava/app/src/main/java/com/rokid/mobile/sdk/demo/java/fragment/BindFragment.java
package com.rokid.mobile.sdk.demo.java.fragment;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.TextInputEditText;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.LinearLayout;

import com.rokid.mobile.lib.base.util.Logger;
import com.rokid.mobile.lib.entity.bean.bluetooth.BTDeviceBean;
import com.rokid.mobile.lib.entity.bean.wifi.WifiBean;
import com.rokid.mobile.sdk.demo.java.R;
import com.rokid.mobile.sdk.demo.java.adapter.item.BleItem;
import com.rokid.mobile.sdk.demo.java.presenter.BindFragmentPresenter;
import com.rokid.mobile.sdk.demo.java.util.SoftKeyBoardUtil;
import com.rokid.mobile.sdk.demo.java.util.WifiEngine;
import com.rokid.mobile.ui.recyclerview.adapter.BaseRVAdapter;

import butterknife.BindView;

/**
 * Created by tt on 2018/2/24.
 */

public class BindFragment extends BaseFragment<BindFragmentPresenter> {

    @BindView(R.id.fragment_bind_ble_name_edit)
    TextInputEditText bleNameTxt;

    @BindView(R.id.fragment_bind_scan_ble_btn)
    Button scanBleBtn;

    @BindView(R.id.fragment_bind_scan_ble_load_layer)
    LinearLayout loadLayer;

    @BindView(R.id.fragment_bind_ble_list_rv)
    RecyclerView bleRV;

    @BindView(R.id.fragment_bind_wifi_name_edit)
    TextInputEditText wifiNameTxt;

    @BindView(R.id.fragment_bind_wifi_pwd_edit)
    TextInputEditText wifiPwdTxt;

    @BindView(R.id.fragment_bind_btn)
    Button bindBtn;

    private BaseRVAdapter<BleItem> mAdapter;

    private BleItem currentItem;

    private WifiBean wifiBean;

    @Override
    protected int getLayoutId() {
        return R.layout.fragment_bind;
    }

    @Override
    protected BindFragmentPresenter initPresenter() {
        return new BindFragmentPresenter(this);
    }

    @Override
    protected void initVariables(View rootView, ViewGroup container, @Nullable Bundle savedInstanceState) {
        bleNameTxt.setSelection(bleNameTxt.getText().length());

        mAdapter = new BaseRVAdapter<>();

        bleRV.setLayoutManager(new LinearLayoutManager(getActivity()));
        bleRV.setAdapter(mAdapter);
    }

    @Override
    protected void initListeners() {
        scanBleBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (TextUtils.isEmpty(bleNameTxt.getText())) {
                    Logger.d("ble name is empty");
                    showToastShort(getString(R.string.fragment_bind_ble_name_empty));
                    return;
                }

                SoftKeyBoardUtil.hideSoftKeyboard(getContext(), scanBleBtn);
                loadLayer.setVisibility(View.VISIBLE);
                bleRV.setVisibility(View.GONE);
                getPresenter().stopScanBle();
                mAdapter.clearAllItemView();
                getPresenter().startScanBle(bleNameTxt.getText().toString());
            }
        });

        mAdapter.setOnItemViewClickListener(new BaseRVAdapter.OnItemViewClickListener<BleItem>() {
            @Override
            public void onItemViewClick(BleItem bleItem, int i1, int i2) {
                if (bleItem == null) {
                    return;
                }

                if (bleItem.getData() == null) {
                    return;
                }

                currentItem = bleItem;
                currentItem.hideBindTxt();
                bleItem.showProgressBar();
                getPresenter().connectBle(bleItem.getData().getName());
            }
        });

        bindBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (TextUtils.isEmpty(wifiNameTxt.getText()) &&
                        TextUtils.isEmpty(wifiPwdTxt.getText())) {
                    showToastShort(getString(R.string.fragment_bind_wifi_empty));
                    return;
                }

                SoftKeyBoardUtil.hideSoftKeyboard(getContext(), bindBtn);
                getPresenter().sendBindData(wifiNameTxt.getText().toString(),
                        wifiPwdTxt.getText().toString(),
                        wifiBean.getBssid());
            }
        });

    }

    @Override
    public void onResume() {
        super.onResume();
        initWifiInfo();
    }

    @Override
    public void setUserVisibleHint(boolean isVisibleToUser) {
        super.setUserVisibleHint(isVisibleToUser);
        if (!isVisibleToUser && null != getPresenter()) {
            getPresenter().stopScanBle();
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (null != getPresenter()) {
            getPresenter().stopScanBle();
        }
    }

    /**
     * 初始化wifi信息
     */
    private void initWifiInfo() {
        wifiBean = WifiEngine.getInstance().getConnectWifiInfo();
        if (null == wifiBean) {
            return;
        }

        if (!TextUtils.isEmpty(wifiBean.getSsid())) {
            wifiNameTxt.setText(wifiBean.getSsid());
            wifiNameTxt.setSelection(wifiBean.getSsid().length());
        }
    }

    public void setBleData(BTDeviceBean bleData) {
        hideLoadLayer();

        if (null == bleData) {
            showToastShort(getString(R.string.fragment_bind_ble_list_empty));
            return;
        }

        bleRV.setVisibility(View.VISIBLE);
        mAdapter.addItemView(new BleItem(bleData));
    }

    /**
     * 连接蓝牙成功
     */
    public void connectBleSuccess() {
        getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                currentItem.hideProgressBar();
                currentItem.showBindTxt();
            }
        });

        showToastShort(getString(R.string.fragment_bind_ble_connect_success));
    }

    /**
     * 连接蓝牙失败
     */
    public void connectBleFail(String text) {
        getActivity().runOnUiThread(new Runnable() {
            @Override
            public void run() {
                currentItem.hideProgressBar();
            }
        });
        showToastShort(text);
    }

    public void hideLoadLayer() {
        loadLayer.setVisibility(View.GONE);
    }
}

File: RokidSDKDemokotlin/app/src/main/kotlin/com/rokid/mobile/sdk/demo/fragment/BindFragment.kt
package com.rokid.mobile.sdk.demo.fragment

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import android.widget.Toast
import com.rokid.lib_appbase.recyclerview.adapter.BaseRVAdapter
import com.rokid.mobile.lib.base.util.Logger
import com.rokid.mobile.lib.entity.bean.bluetooth.BTDeviceBean
import com.rokid.mobile.lib.entity.bean.bluetooth.DeviceBinderData
import com.rokid.mobile.lib.entity.bean.wifi.WifiBean
import com.rokid.mobile.lib.xbase.binder.bluetooth.callBack.IBTConnectCallBack
import com.rokid.mobile.lib.xbase.binder.bluetooth.callBack.IBTSendCallBack
import com.rokid.mobile.lib.xbase.binder.bluetooth.exception.BleException
import com.rokid.mobile.sdk.RokidMobileSDK
import com.rokid.mobile.sdk.demo.R
import com.rokid.mobile.sdk.demo.adapter.item.BleItem
import com.rokid.mobilesdkdemo.util.WifiEngine
import kotlinx.android.synthetic.main.fragment_bind.view.*

/**
 *
 *配网模块
 * Created by wangshuwen on 2017/12/4.
 */
class BindFragment : Fragment() {

    private var rootView: View? = null
    /**
     * 显示蓝牙设备的RecyclerView
     */
    private lateinit var rv: RecyclerView
    /**
     * 扫描蓝牙的btn
     */
    private lateinit var scanBt: Button
    /**
     * 蓝牙前缀的输入框
     */
    private lateinit var bleNameEt: EditText
    /**
     * 绑定的btn
     */
    private lateinit var sendDataBtn: Button
    /**
     * wifi密码的输入框
     */
    private lateinit var wifiPasswordEdit: EditText
    /**
     * wifi名字的输入框
     */
    private lateinit var wifiSSIDEdit: EditText
    /**
     * wifi信息
     */
    private var wifiInfo: WifiBean? = null
    /**
     * 当前连接的蓝牙item
     */
    private var currentBleItem: BleItem? = null

    private lateinit var mAdapter: BaseRVAdapter<BleItem>

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        if (null == rootView) {
            rootView = inflater!!.inflate(R.layout.fragment_bind, container!!, false)
            initView()
            initListener()
        }
        return rootView
    }

    override fun onResume() {
        super.onResume()
        initWifiInfo()
    }

    private fun initView() {
        scanBt = rootView!!.fragment_bind_scan

        bleNameEt = rootView!!.fragment_bind_blename_et
        bleNameEt.setSelection(bleNameEt.length())

        wifiPasswordEdit = rootView!!.fragment_bind_wifipwd_txt
        wifiSSIDEdit = rootView!!.fragment_bind_wifiname_txt
        sendDataBtn = rootView!!.fragment_bind_send_data_btn

        rv = rootView!!.fragment_bind_rv
        rv.apply {
            layoutManager = LinearLayoutManager(activity)
            mAdapter = BaseRVAdapter()
            adapter = mAdapter

        }

    }

    private fun initListener() {
        //扫描按钮
        scanBt.setOnClickListener {
            val nameMark = bleNameEt.text.toString()
            if (nameMark.isEmpty()) {
                Logger.e("BLE name mark is empty please edit")
                Toast.makeText(this@BindFragment.activity, "请输入蓝牙前缀 ~~~", Toast.LENGTH_SHORT).show()
                return@setOnClickListener
            }
            //停止扫描
            RokidMobileSDK.binder.stopBTScanAndClearList()
            mAdapter.clearAllItemView()
            doStartScanBle(nameMark)
        }
        //蓝牙设备点击
        mAdapter.setOnItemViewClickListener { bleItem, sectionKey, sectionItemPosition ->
            Logger.i("onItemClick position=" + sectionItemPosition + " blueToothName=" + bleItem.data.name)
            stopScan()
            currentBleItem = bleItem
            bleItem.showProgress()
            connectBle(bleItem.data.name)
        }
        //发送绑定数据 btn
        sendDataBtn.setOnClickListener {
            sendBindData()
        }

    }


    /**
     * 初始化wifi信息
     */
    private fun initWifiInfo() {
        Logger.i("initWifiInfo is called")
        wifiInfo = WifiEngine.connectWifiInfo

        wifiInfo?.let {
            val ssid = wifiInfo!!.getSsid()
            wifiSSIDEdit.setText(ssid)
            wifiSSIDEdit.setSelection(wifiSSIDEdit.length())
        }

    }

    /**
     * 扫描蓝牙
     */
    fun doStartScanBle(nameMark: String) {
        RokidMobileSDK.binder.startBTScan(nameMark) { bleDevice ->
            bleDevice?.let {
                mAdapter.addItemView(BleItem(bleDevice))
            }
        }
    }

    /**
     * 停止扫描
     */
    fun stopScan() {
        RokidMobileSDK.binder.stopBTScan()
    }


    /**
     * 连接蓝牙
     */
    fun connectBle(bleName: String) {
        RokidMobileSDK.binder.connectBT(bleName, object : IBTConnectCallBack {

            override fun onConnectSucceed(p0: BTDeviceBean?) {
                this@BindFragment.activity.runOnUiThread {
                    Toast.makeText(this@BindFragment.activity, "连接成功 请输入wifi信息", Toast.LENGTH_LONG).show()
                    currentBleItem?.hideProgress()
                }
            }

            override fun onConnectFailed(p0: BTDeviceBean?, p1: BleException?) {
                this@BindFragment.activity.runOnUiThread {
                    Toast.makeText(this@BindFragment.activity,
                            "连接失败 errorCode=" + (p0 ?: "") + "errorMsg= " + (p1 ?: ""),
                            Toast.LENGTH_LONG).show()
                    currentBleItem?.hideProgress()
                }
            }

        })
    }


    fun sendBindData() {
        //todo 输入框
        val userId = ""
        val bssid = wifiInfo?.bssid
        val ssid = wifiSSIDEdit.text.toString()
        val pwd = wifiPasswordEdit.text.toString()

        if (TextUtils.isEmpty(ssid) && TextUtils.isEmpty(pwd)) {
            Toast.makeText(this@BindFragment.activity, "wifi名字和密码为空 请重新输入", Toast.LENGTH_LONG).show()
            return
        }
        if (TextUtils.isEmpty(userId)) {
            Logger.e("send bind data userId is empty")
        }
        val bindData = DeviceBinderData.newBuilder()
                .userId(userId)
                .wifiPwd(pwd)
                .wifiSsid(ssid)
                .wifiBssid(bssid)
                .build()


        RokidMobileSDK.binder.sendBTBinderData(bindData, object : IBTSendCallBack {
            override fun onSendSuccess(p0: BTDeviceBean?) {
                Logger.e("Send Success -----------------")
                this@BindFragment.activity.runOnUiThread {
                    Toast.makeText(this@BindFragment.activity, "发送数据成功", Toast.LENGTH_LONG).show()

                }
            }

            override fun onSendFailed(p0: BTDeviceBean?, p1: BleException?) {
                Logger.e("SendFailed -----------------")
                this@BindFragment.activity.runOnUiThread {
                    Toast.makeText(this@BindFragment.activity,
                            "发送数据失败 errorCode=" + (p0 ?: "") + "errorMsg= " + (p1 ?: ""),
                            Toast.LENGTH_LONG).show()
                }
            }
        })
    }


}